src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/SamplingLoader.hpp>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/pool.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <cast_core/helpers/pipeline_helpers.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/PipelineGenerator.hpp>
+
+namespace genny::actor {
+
+/** @private */
+using index_type = std::tuple<DocumentGenerator, std::optional<DocumentGenerator>, std::string>;
+using namespace bsoncxx;
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+
+/** @private */
+struct SamplingLoader::PhaseConfig {
+    PhaseConfig(PhaseContext& context, mongocxx::pool::entry& client, ActorId id)
+        : database{(*client)[context["Database"].to<std::string>()]},
+          collection{database[context["Collection"].to<std::string>()]},
+          insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
+          numBatches{context["Batches"].to<IntegerSpec>()},
+          // A sample size is optional. If unspecified, we'll make a sample size such that each
+          // sample document is inserted one time.
+          sampleSize{
+              context["SampleSize"].maybe<IntegerSpec>().value_or(numBatches * insertBatchSize)},
+          pipelineSuffixGenerator{context["Pipeline"]
+                                      .maybe<PipelineGenerator>(context, id)
+                                      .value_or(PipelineGenerator{})} {}
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    // User can specify any pipeline here, but it's worth noting that some will not play nicely. For
```

Comment:
Now, that we have extended documentation in the SamplingLoader.yml I'd just direct the readers to that file for details (would probably do that in the line above `insertBatchSize`)

Reply:
ping?

Reply:
Done!